hbase-server/src/main/java/org/apache/hadoop/hbase/util/DirectMemoryUtils.java

```diff
@@ -77,36 +76,16 @@ public class DirectMemoryUtils {
     HAS_MEMORY_USED_ATTRIBUTE = a != null;
   }
 
-  /**
-   * @return the setting of -XX:MaxDirectMemorySize as a long. Returns 0 if -XX:MaxDirectMemorySize
-   *         is not set.
-   */
+  /** Returns the direct memory limit of the current progress */
   public static long getDirectMemorySize() {
-    RuntimeMXBean runtimemxBean = ManagementFactory.getRuntimeMXBean();
-    List<String> arguments = runtimemxBean.getInputArguments();
-    long multiplier = 1; // for the byte case.
-    for (String s : arguments) {
-      if (s.contains("-XX:MaxDirectMemorySize=")) {
-        String memSize = s.toLowerCase(Locale.ROOT).replace("-xx:maxdirectmemorysize=", "").trim();
-
-        if (memSize.contains("k")) {
-          multiplier = 1024;
-        }
-
-        else if (memSize.contains("m")) {
-          multiplier = 1048576;
-        }
-
-        else if (memSize.contains("g")) {
-          multiplier = 1073741824;
-        }
-        memSize = memSize.replaceAll("[^\\d]", "");
-
-        long retValue = Long.parseLong(memSize);
-        return retValue * multiplier;
-      }
+    try {
+      Field directMemoryLimit = PlatformDependent.class.getDeclaredField("MAX_DIRECT_MEMORY");
```

Comment:
PlatformDependent.maxDirectMemory is enough?

https://github.com/netty/netty/blob/cf670fab75ad3c77d6b37883104b259894d4bd71/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L213

Reply:
Thanks for reviewing Duo @Apache9 

I think maybe we cannot. Here is the reason:
1.  in the master branch, the hbase-thirdparty version is 4.1.2
https://github.com/apache/hbase/blob/master/pom.xml#L870

2. the the hbase-thirdparty 4.1.2, the netty version is 4.1.82.Final
https://github.com/apache/hbase-thirdparty/blob/230e5ebc432c2004b4ab744dca6627697f9f2a76/pom.xml#L135

3. in the netty 4.1.82.Final, the return value is DIRECT_MEMORY_LIMIT
https://github.com/netty/netty/blob/47799635143d7a11b56c4a4e9a1e65ca221d28ca/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L407

however, the value can be overrided if org.apache.hbase.thirdparty.io.netty.maxDirectMemory is set.
https://github.com/netty/netty/blob/47799635143d7a11b56c4a4e9a1e65ca221d28ca/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L179

So I think this API is not stable enough. That's the reason why I give up this.

Reply:
OK, on branch 4.1 it returns the netty limit. Should we provide a fallback way if netty one is avaiable?

Reply:
We use reflection to get this field so it will not trigger a compilation error if netty changes this field in the future, so we can not find it when upgrading netty.

The estimateMaxDirectMemory method is public, so maybe we could just call it and store the return value by our own?

Reply:
Oh, in the old version, this method is private. Didn't notice it has been public now. Thanks for point it out.

You are right，this is better than reflection. I'll fix this. Thanks Duo.@Apache9